vm/actor/src/builtin/miner/partition_state.rs

```diff
@@ -0,0 +1,730 @@
+        if !self.recoveries.is_empty() {
+            return Err("unexpected recoveries while processing expirations".into());
+        }
+        if !self.recovering_power.is_zero() {
+            return Err("unexpected recovering power while processing expirations".into());
+        }
+
+        // Nothing expiring now should have already terminated.
+        if self.terminated.contains_any(&expired_sectors) {
+            return Err("expiring sectors already terminated".into());
+        }
+
+        // Mark the sectors as terminated and subtract sector power.
+        self.terminated |= &expired_sectors;
+        self.faults -= &expired_sectors;
+        self.live_power -= &(&popped.active_power + &popped.faulty_power);
+        self.faulty_power -= &popped.faulty_power;
+
+        // Record the epoch of any sectors expiring early, for termination fee calculation later.
+        self.record_early_termination(store, until, &popped.early_sectors)
+            .map_err(|e| ActorError::downcast_wrap(e, "failed to record early terminations"))?;
+
+        Ok(popped)
+    }
+
+    /// Marks all non-faulty sectors in the partition as faulty and clears recoveries, updating power memos appropriately.
+    /// All sectors' expirations are rescheduled to the fault expiration, as "early" (if not expiring earlier)
+    /// Returns the power of the newly faulty and failed recovery sectors.
+    pub fn record_missed_post<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        fault_expiration: ChainEpoch,
+        quant: QuantSpec,
+    ) -> Result<(PowerPair, PowerPair), Box<dyn StdError>> {
+        // Collapse tail of queue into the last entry, and mark all power faulty.
+        // Load expiration queue
+        let mut queue = ExpirationQueue::new(store, &self.expirations_epochs, quant)
+            .map_err(|e| format!("failed to load partition queue: {:?}", e))?;
+
+        queue
+            .reschedule_all_as_faults(fault_expiration)
+            .map_err(|e| ActorError::downcast_wrap(e, "failed to reschedule all as faults"))?;
+
+        // Save expiration queue
+        self.expirations_epochs = queue.amt.flush()?;
+
+        // Compute faulty power for penalization. New faulty power is the total power minus already faulty.
+        let new_fault_power = &self.live_power - &self.faulty_power;
+        let failed_recovery_power = self.recovering_power.clone();
+
+        // Update partition metadata
+        let all_faults = self.live_sectors();
+        self.faults = all_faults;
+        self.recoveries = BitField::new();
+        self.faulty_power = self.live_power.clone();
+        self.recovering_power = PowerPair::zero();
+
+        Ok((new_fault_power, failed_recovery_power))
+    }
+
+    pub fn pop_early_terminations<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        max_sectors: u64,
+    ) -> Result<(TerminationResult, /* has more */ bool), Box<dyn StdError>> {
```

Comment:
Instead of labeling the return value, might be better to doc comment this method as best as you can

Reply:
This happens many places throughout too. I'd say to have them as part of the doc comments

Reply:
In many cases I genuinely don't know exactly what the return value means, this seemed like the most direct way to translate their named return values. I'm not disagreeing but I don't think I have the understanding to write more meaningful doc comments than they already have